beacon_node/client/src/address_change_broadcast.rs

```diff
@@ -0,0 +1,118 @@
+use crate::*;
+use lighthouse_network::PubsubMessage;
+use network::NetworkMessage;
+use slog::{debug, info, Logger};
+use slot_clock::SlotClock;
+use std::cmp;
+use std::time::Duration;
+use tokio::sync::mpsc::UnboundedSender;
+use tokio::time::sleep;
+use types::EthSpec;
+
+/// The size of each chunk of addresses changes to be broadcast at the Capella
+/// fork.
+const BROADCAST_CHUNK_SIZE: usize = 128;
+/// The delay between broadcasting each chunk.
+const BROADCAST_CHUNK_DELAY: Duration = Duration::from_millis(500);
+
+/// Waits until the Capella fork epoch and then publishes any bls to execution
+/// address changes which were placed in the pool prior to the fork.
+///
+/// Does nothing if the Capella fork has already happened.
+///
+/// Address changes are published in chunks, with a delay between each chunk.
+/// This helps reduce the load on the P2P network and also helps prevent us from
+/// clogging our `network_send` channel and being late to publish
+/// blocks, attestations, etc.
+pub async fn broadcast_address_changes_at_capella<T: BeaconChainTypes>(
+    chain: &BeaconChain<T>,
+    network_send: UnboundedSender<NetworkMessage<T::EthSpec>>,
+    log: &Logger,
+) {
+    let spec = &chain.spec;
+    let slot_clock = &chain.slot_clock;
+
+    let capella_fork_slot = if let Some(epoch) = spec.capella_fork_epoch {
+        epoch.start_slot(T::EthSpec::slots_per_epoch())
+    } else {
+        // Exit now if Capella is not defined.
+        return;
+    };
+
+    // Wait until the Capella fork epoch.
+    loop {
+        match slot_clock.duration_to_slot(capella_fork_slot) {
```

Comment:
Seeing as this function only runs once and _only_ if it lands at the fork exactly, do we need to store the broadcast indices on disk? I was thinking we could just snapshot the contents of the pool at the fork boundary, store that in-memory in this function and then call it a day once that snapshot is fully broadcast.

If the user restarts their node during this process they're going to miss out with either of these approaches. If we wanted to strengthen the guarantee to account for nodes that go offline we should allow broadcasting the pool contents at any point after Capella, if we have messages on-disk that are awaiting broadcast.

So depending on what we decide:

- Option 1: broadcast at any time after fork: need the indices on-disk.
- Option 2: broadcast only at the fork slot: no need to store indices on-disk, no need for a schema migration.

Reply:
> Seeing as this function only runs once and only if it lands at the fork exactly

This is all still WIP, so I wouldn't put too much weight on the peculiarities of the timing here. I'll give it all another run-through and try and get it to be sane.

> we should allow broadcasting the pool contents at any point after Capella

This is something @potuz is working on over in https://github.com/prysmaticlabs/prysm/pull/11871. I think it's an useful idea and I've been holding off on adding this functionality until that PR is merged. I figure we can try and match the URL.

Reply:
Now that I've been tagged, the reason we haven't merged it (or that I haven't pushed back the resistance to have custom endpoints) is that a user can simply restart and resubmit his changes to the endpoint. I do think there's some minimal benefit in allowing a user to resubmit everything including all messages that were obtained in gossip so that less messages are lost by dropping

Reply:
Alrighty, I've implemented the following:

- Indices requiring Capella broadcast are stored on disk.
- The broadcast routine runs both at the Capella fork and each time the node starts.
- After each address change is broadcast, the relevant val index is removed from the list.
- There is no way to re-trigger the broadcast of address changes.

This provides *some* guarantee that any pre-Capella address changes will eventually be sent to our networking service, regardless of reboots. I'm sure there's some edge-cases where we don't manage to actually get a message onto the network, however I don't think we need to cover these scenarios.

I've avoided the broadcast re-triggering logic since I think it's scope creep on this PR. We can add it later if we want it.